src/pv.rs

```diff
@@ -43,28 +44,118 @@ fn packet_dump(packet: &PvPacket) {
 }
 
 #[derive(Debug)]
-pub struct PvPacket {
+pub struct Packet {
     pub start: u32,       // payload offset pointing the start of payload. ex. 256
     pub end: u32,         // payload offset point the end of payload. ex. 1000
-    pub size: u32,        // total size of buffer. ex. 2048
+    pub buffer_size: u32, // total size of buffer. ex. 2048
     pub buffer: *mut u8,  // buffer address.
     private: *mut c_void, // DO NOT touch this.
 }
 
-impl PvPacket {
-    fn new() -> PvPacket {
-        PvPacket {
+impl Packet {
+    fn new() -> Packet {
+        Packet {
             start: 0,
             end: 0,
-            size: 0,
+            buffer_size: 0,
             buffer: std::ptr::null_mut(),
             private: std::ptr::null_mut(),
         }
     }
+
+    /* check whether the new header can be attached to payload from the start of packet
+    add header to payload and return Ok if it's possible, or return Err */
+    pub fn attach_header(&mut self, header: Vec<u8>) -> Result<(), ()> {
+        if self.start >= header.len() as u32 {
+            // header can be attached in empty space of headroom
+
+            self.start = self.start - header.len() as u32;
+
+            unsafe {
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+
+            Ok(())
+        } else if header.len() as u32 <= self.buffer_size - self.end + self.start {
+            // header can't be attached in headroom but can be by shifting payload
+            unsafe {
+                // move payload (memmove)
+                copy(
+                    self.buffer.offset(self.start as isize),
+                    self.buffer.offset(header.len() as isize),
+                    (self.end - self.start) as usize,
+                );
+
+                // set start and end of data
+                self.start = 0;
+                self.end = self.end + (header.len() as u32 - self.start);
```

Comment:
Very very hard to understand
Why not save self.end - self.start as payload size and use it here and memmove?
`self.end = header.len + payload_size`

Reply:
the function is removed because of it's not used